start chord header tasks as soon as possible

mjhoffman65 opened the pull request:
Fixes #3021 . Previously, header tasks were not submitted to the broker for
execution until the entire generator was consumed.

Continuation of https://github.com/celery/celery/pull/6112 and much of this was originally pulled from https://github.com/celery/celery/pull/3043.

Update: this was split into 2 PRs per below review: https://github.com/celery/celery/pull/6589 should be merged first.  This has been rebased on top of that branch.

auvipy commented on celery/canvas.py.1:
```diff
@@ -0,0 +1,1806 @@
+"""Composing task work-flows.
+
+.. seealso:
+
```
curious what this file doing? just renamed?

mjhoffman65 replied:
Oops, this shouldn't have been included.  It's should be gone now.

auvipy reviewed (changes requested):
I must have to say great effort so far, could you check this PR https://github.com/celery/celery/pull/6542 and split the related part to another PR and after that get merged rebase this and review & test?

mjhoffman65 commented:
> I must have to say great effort so far, could you check this PR #6542 and split the related part to another PR and after that get merged rebase this and review & test?

Thanks!  Just to make sure I understand, split this PR into 2: first one will replace #6542 and include only the `_regen` changes and the second one will include the rest?

auvipy commented:
> > I must have to say great effort so far, could you check this PR #6542 and split the related part to another PR and after that get merged rebase this and review & test?
> 
> @auvipy Thanks! Just to make sure I understand, split this PR into 2: first one will replace #6542 and include only the `_regen` changes and the second one will include the rest?

yes.

auvipy commented:
also, check the conflicts please

thedrow commented:
I think we had another PR for this one, didn't we? @auvipy 

@mjhoffman65 Thank you very much for the contribution.
Would you kindly rebase this so I can review?

mjhoffman65 commented:
@auvipy @thedrow Merge conflicts have been resolved and this PR has been split (https://github.com/celery/celery/pull/6589) as requested.

thedrow commented on celery/utils/functional.py:
```diff
@@ -190,22 +204,40 @@ def __length_hint__(self):
+            return True
+
+        try:
+            next(iter(self))
```
Isn't this line going to be a problem?
We'll lose an item while checking if it exists.
Shouldn't we also add it to `self. __consumed`?

mjhoffman65 replied:
Added a test to make sure we aren't losing any items: https://github.com/celery/celery/pull/6576/commits/f0d4f2e8915dd5b2710c372c5454218ab89e08db.  That next call drops into `self.__iter__` which will add to `self.__consumed`.

thedrow reviewed (changes requested):
I need to review this in detail.
In the meantime, I suspect one thing has to be changed.

auvipy commented:
rebase please

mjhoffman65 commented:
> rebase please

:+1: rebased

maybe-sybr commented:
A quick look at the changes to the redis backend make me think that we can't land this in a patch release since it introduces new key dependencies when handling chords in flight.

I'm yet to review in full.

thedrow commented:
@maybe-sybr Feel free to move the milestone to 5.1.0 if you deem it necessary.

auvipy commented on celery/canvas.py:
```diff
@@ -165,7 +160,6 @@ def __init__(self, task=None, args=None, kwargs=None, options=None,
                 options=dict(options or {}, **ex),
                 subtask_type=subtask_type,
                 immutable=immutable,
-                chord_size=None,
```
do the changes related to chord_size going to break the current API?

mjhoffman65 replied:
I don't think it should break anything and the tests are passing for me.  Is there a missing test case I should add to make sure the API is ok?

thedrow replied:
If the integration tests will pass, everything is fine and this can be removed. Can you run them to make sure?

mjhoffman65 replied:
Ran the integration tests with docker and I am getting the same results on this branch as I do on the master branch (1 failure that is unrelated to these changes).

![image](https://user-images.githubusercontent.com/1480930/109672180-c51cf780-7b42-11eb-9e09-a0cbe30237b1.png)

auvipy commented on celery/canvas.py:
```diff
@@ -1179,12 +1166,22 @@ def _apply_tasks(self, tasks, producer=None, app=None, p=None,
+                sig, res, group_id = current_task
+                _chord = sig.options.get("chord") or chord
+                if _chord is not None and next_task is None:
+                    chord_length = task_index + 1
```
so we are going to use chord_length instead of chord_size?

thedrow replied:
Does chord_size appear anywhere else? I remember that it does but I'm not sure.
Let's try to keep the same terminology.

mjhoffman65 replied:
Changed `chord_length` back to `chord_size` to keep the same terminology.

auvipy commented on celery/canvas.py:
```diff
@@ -265,7 +259,6 @@ def clone(self, args=None, kwargs=None, **opts):
                                          'kwargs': kwargs,
                                          'options': deepcopy(opts),
                                          'subtask_type': self.subtask_type,
-                                         'chord_size': self.chord_size,
```
@thedrow here is another

auvipy commented on celery/canvas.py:
```diff
@@ -530,8 +523,6 @@ def _apply_async(self):
     kwargs = getitem_property('kwargs', 'Keyword arguments to task.')
     options = getitem_property('options', 'Task execution options.')
     subtask_type = getitem_property('subtask_type', 'Type of signature')
-    chord_size = getitem_property(
```
are we spliting chord_size with chord_length & chord_size_bytes?

mjhoffman65 replied:
Removed the references to `chord_length` to avoid confusion.  Here `chord_size` is removed from the chord object since it is set on the redis backend instead.  

`chord_size_bytes` exists because redis returns a integer back as bytes.
See https://github.com/andymccurdy/redis-py/blob/1a41cfd95a53a95b078084d8627be6b6fba3bb71/tests/test_commands.py#L219

auvipy replied:
cool thanks for clearing

auvipy commented:
I was working on fixing the CI and running unit test got this error in CI

t/unit/utils/test_functional.py::test_regen::test_nonzero__does_not_consume_more_than_first_item FAILED [ 66%]

=================================== FAILURES ===================================
________ test_regen.test_nonzero__does_not_consume_more_than_first_item ________

self = <t.unit.utils.test_functional.test_regen object at 0x7f7e18c6a860>

>   ???
E   assert not True
E    +  where True = <t.unit.utils.test_functional.test_regen object at 0x7f7e18c6a860>.consumed_second_item

t/unit/utils/test_functional.py:145: AssertionError

mjhoffman65 commented:
> I was working on fixing the CI and running unit test got this error in CI
> 
> t/unit/utils/test_functional.py::test_regen::test_nonzero__does_not_consume_more_than_first_item FAILED [ 66%]
> 
> =================================== FAILURES ===================================
> ________ test_regen.test_nonzero__does_not_consume_more_than_first_item ________
> 
> self = <t.unit.utils.test_functional.test_regen object at 0x7f7e18c6a860>
> 
> > ???
> > E   assert not True
> > E    +  where True = <t.unit.utils.test_functional.test_regen object at 0x7f7e18c6a860>.consumed_second_item
> 
> t/unit/utils/test_functional.py:145: AssertionError

Which branch are you seeing this test failure on?  I'm having trouble recreating it

mjhoffman65 commented:
(Rebased off of master)

auvipy commented:
https://github.com/celery/celery/pull/6587/checks?check_run_id=1749550836

mjhoffman65 commented:
> https://github.com/celery/celery/pull/6587/checks?check_run_id=1749550836

Thanks!  Looking into it

auvipy commented:
pull from master again plz

mjhoffman65 commented:
```
_______________ ERROR collecting t/unit/utils/test_functional.py _______________
ImportError while importing test module '/home/runner/work/celery/celery/t/unit/utils/test_functional.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.6.12/x64/lib/python3.6/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
t/unit/utils/test_functional.py:2: in <module>
    from celery.utils.functional import (DummyContext, first, firstmethod,
E   ImportError: cannot import name 'lookahead'
=========================== short test summary info ============================
ERROR t/unit/utils/test_functional.py
!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!
========================= 11 skipped, 1 error in 4.63s =========================
```
@auvipy The import error on the tests makes me think these checks might not be using my branch?  The `lookahead` was introduced in this branch.

auvipy commented:
because that is not yet merged into master branch yet :D

auvipy commented:
ops sorry that looks weired

auvipy reviewed (changes requested):
can you send this unit test fix to a separate PR to check the CI?

mjhoffman65 commented:
@auvipy https://github.com/celery/celery/pull/6613

auvipy commented on celery/canvas.py:
```diff
@@ -1055,9 +1044,7 @@ def from_dict(cls, d, app=None):
-        return _upgrade(
-            d, group(rebuilt_tasks, app=app, **d['options']),
-        )
+        return group(rebuilt_tasks, app=app, **d['options'])
```
can we add some comment here explaining the change?

mjhoffman65 replied:
`_upgrade` used to just set the chord size which has been refactored out since it is set on the backend now instead.
![image](https://user-images.githubusercontent.com/1480930/108213570-2e414b80-70fd-11eb-955e-bf76320a3d99.png)

auvipy replied:
btw do you have the bandwidth to check the pypy failures? maybe in a different PR. also as we don't have CI for Integration tests now, did all the integration tests pass locally?

mjhoffman65 replied:
yep, I can take a look at the pypy failures in a separate PR.  Integration tests are all passing locally

mjhoffman65 replied:
Here's a PR I started to look into the PyPy failures: https://github.com/celery/celery/pull/6635.  Need some feedback on it when you get some time @auvipy.

mjhoffman65 commented:
Is there anything I can do to move this forward?  I'm happy to help out with the build separately, but also eager to get these changes into master.

auvipy commented:
> Is there anything I can do to move this forward? I'm happy to help out with the build separately, but also eager to get these changes into master.

this will surely go into celery 5.1, we just have to see if we are going to release 5.0.6 before that.

thedrow commented:
We're not going to release 5.0.6. Instead this will land in 5.1.0.

thedrow commented on celery/canvas.py:
```diff
@@ -1179,12 +1166,22 @@ def _apply_tasks(self, tasks, producer=None, app=None, p=None,
         app = app or self.app
         with app.producer_or_acquire(producer) as producer:
-            for sig, res in tasks:
+            for task_index, (current_task, next_task) in enumerate(
```
We should probably add comments here.

mjhoffman65 replied:
Added a comment here.

thedrow commented on celery/canvas.py:
```diff
@@ -1221,20 +1218,43 @@ def freeze(self, _id=None, group_id=None, chord=None,
-        if isinstance(self.tasks, MutableSequence):
-            self.tasks[:] = new_tasks
+        if isinstance(self.tasks, _regen):
+            tasks1, tasks2 = itertools.tee(self._unroll_tasks(self.tasks))
```
Please add a comment explaining that we're draining from a generator.

mjhoffman65 replied:
Added a comment.

thedrow commented on celery/canvas.py:
```diff
@@ -1221,20 +1218,43 @@ def freeze(self, _id=None, group_id=None, chord=None,
     _freeze = freeze
 
+    def _freeze_tasks(self, tasks, group_id, chord, root_id, parent_id):
+        for group_index, task in enumerate(tasks):
+            yield task.freeze(group_id=group_id,
+                              chord=chord,
+                              root_id=root_id,
+                              parent_id=parent_id,
+                              group_index=group_index)
```
```suggestion
            yield from (task.freeze(group_id=group_id,
                              chord=chord,
                              root_id=root_id,
                              parent_id=parent_id,
                              group_index=group_index)
                              for group_index, task in enumerate(tasks))
```
Isn't this neater?

thedrow commented on celery/canvas.py:
```diff
@@ -1221,20 +1218,43 @@ def freeze(self, _id=None, group_id=None, chord=None,
+                              group_index=group_index)
+
+    def _unroll_tasks(self, tasks):
+        for task in tasks:
+            task = maybe_signature(task, app=self._app).clone()
+            yield task
```
```suggestion
       yield from (maybe_signature(task, app=self._app).clone() for task in tasks)
```

Isn't this neater?

thedrow commented on celery/canvas.py:
```diff
@@ -1452,6 +1471,7 @@ def run(self, header, body, partial_args, app=None, interval=None,
         # we execute the body manually here.
         else:
             body.delay([])
+            header_result = self.app.GroupResult(*header_result_args)
```
Was this a different bug?

mjhoffman65 replied:
Creating a GroupResult will drain the header generator, so I had to refactor it out of `apply_chord`. This line is the last piece of that refactoring.

thedrow replied:
Got it.

thedrow commented on celery/utils/functional.py:
```diff
@@ -160,6 +160,19 @@ def uniq(it):
+        [(0, 1), (1, 2), (2, 3), (3, 4), (4, 5), (5, None)]
+    """
+    a, b = tee(it)
+    next(b, None)
```
I just learned next() has a default. Thanks!

thedrow reviewed (changes requested):
Overall, I just have some minor comments.
This is a very good job. Well done and thank you for doing this.

thedrow commented:
@mjhoffman65 If you'd like to contribute more, we can add you to our team.

auvipy commented:
> @mjhoffman65 If you'd like to contribute more, we can add you to our team.

I think it's still to early.

thedrow reviewed (changes requested):
Can you please take a look into the test failure?

mjhoffman65 commented:
> Can you please take a look into the test failure?

Should be fixed now

mjhoffman65 commented:
> > @mjhoffman65 If you'd like to contribute more, we can add you to our team.
> 
> I think it's still too early.

I'd love to contribute more.  No worries about joining the team if it is too early.  But let me know what I can help on next.

thedrow reviewed (approved):
LGTM. Thank you!

auvipy commented:
thanks a lot to all of you who were involved to make this PR merge <3

naomielst commented on celery/canvas.py:
```diff
@@ -1221,20 +1222,42 @@ def freeze(self, _id=None, group_id=None, chord=None,
+            # We are draining from a geneator here.
+            tasks1, tasks2 = itertools.tee(self._unroll_tasks(self.tasks))
+            results = regen(self._freeze_tasks(tasks1, group_id, chord, root_id, parent_id))
+            self.tasks = regen(x[0] for x in zip(tasks2, results))
```
Hi @mjhoffman65 , I'm a new project member. 
I am trying to understand what you did here, I would appreciate if you could clarify it for me. 
I don't get why you zipped `tasks2` with `results` only to get `x[0]`, resulting in a generator with only `tasks2`'s tasks. Why did you write this and not `self.tasks = regen(tasks2)`?
TIA :)
